db/db_impl/db_impl_compaction_flush.cc

```diff
@@ -2215,7 +2215,14 @@ Status DBImpl::FlushMemTable(ColumnFamilyData* cfd,
     }
   }
 
-  const bool needs_to_join_write_thread = !entered_write_thread;
+  // Note that, when flush reason is kErrorRecoveryRetryFlush, during the
+  // auto retry resume, we do not care to wait for all pre-existing writes to go
+  // through. In fact, some may be blocked waiting on auto-recovery to clear
+  // memtables in which case joining the write thread here would be a deadlock.
+  // Therefore, write thread will not be joined
+  const bool needs_to_join_write_thread =
+      !entered_write_thread &&
+      flush_reason != FlushReason::kErrorRecoveryRetryFlush;
```

Comment:
Maybe I can refactor this to make more sense. Notice how the below code doesn't even call `SwitchMemtable()` in case of `kErrorRecoveryRetryFlush`, so waiting for pending writes was pointless (?)

Reply:
#11903 for this